rfcs/0076-refined-s3-source.md

```diff
@@ -0,0 +1,88 @@
+
+## List and Fetch
+
+So I am proposing a new solution, which is to use two separate executors to complete the job.
+Let's say they are `list` and `fetch`, aligning the name in [RFC: S3 Source with SQS](https://github.com/risingwavelabs/rfcs/pull/22), 
+and they do the same thing.
+
+* `List` is unique in one source plan. It is responsible for keep listing all files and push the filenames to the downstream.
+* `Fetch` is responsible for fetching the file from S3 and push the file to the downstream, as well as write the received
+  filenames to Hummock.
+
+## How It Works
+
+![img](../assets/refined-s3-source.svg)
+
+### List Part
+
+The `list` cannot push all filenames to the `fetch` part every time. Letting the shuffle operator distribute millions of
+rows every minute is an excessive and unnecessary burden on the system.
+If we do this, the fetch operator must perform `point get` operation on each row to deduplicate, which is also unacceptable for hummock.
+
+The `list` executor will keep listing the files in S3 bucket and try its best to de-duplicate the files (but no guarantee).
+It will record the timestamp when doing the last listing and filter out the files that are created before the timestamp.
+And we can utilize the idea of [RFC: Refine S3 Source: List Objects in A Lazy Way](https://github.com/risingwavelabs/rfcs/pull/74),
+not to list all pages at once, but to list the pages one by one, preventing overload downstream.
+
+`list` executor will keep the max page `p` number as well as the last list timestamp `ts` in its state store.
+When list op is triggered, it filters out the files that are created before `ts` in the first `p` pages.
+If `p` increases (`list` only increase `p` at the request of Meta by barrier), it will push down all items in the `p+1` page
+and update `p` and `ts` in the state store.
+
+### Fetch Part
+
+The `fetch` part will do the ordinary thing, get files by name, read, and yield stream chunks.
+Each `fetch` executor accepts both message stream (filenames from `list`) and barrier stream.
+
+Every `fetch` executor keeps a local task queue, and a state cache in the memory.
+
+* Task queue is used to store the filenames that are not fetched yet.
+  * Task queue is initialized by `select * from state_table where partition in local_vnode`. It just reads all things in the local vnode.
+  * When receiving new filenames from the message stream, it will push the filenames to the task queue.
+* State cache is used to store the filenames that are fetched but not written to hummock yet.
+  * State cache is initialized with an empty HashMap.
+  * When receiving new filenames from the message stream, it will push the filenames to the state cache.
+  * When receiving a barrier, it will write the state cache to hummock and clear the state cache. (please note here we require
+    the hummock not to perform update on insert the same key)
```

Comment:
> please note here we require the hummock not to perform update on insert the same key

Why is this the case? Can you explain more?

Reply:
`list` executor has no grantee on filenames never duplicate.
For example, `file_1` has been done and it has been materialized to the state table. But somehow `list` passed `file_1` again, and `fetch` exec will enqueue `file_1` to the task queue and write it to the state table as a new file when a barrier comes. The default behavior of state table is to overwrite when meeting the same pk and it will reset `file_1` to the init stage.
Or we may do a `point get` every time receive a new file to check whether it has been in the state table.

Reply:
> Or we may do a point get every time receive a new file to check whether it has been in the state table.

+1 for this approach since it is more straight-fordward. Also, I think `fetch` anyway need to do a point get when it receives a file from `list` to determine whether it should be put into the task queue or it should be ignored.